Userland/Libraries/LibGfx/BitmapFont.cpp

```diff
@@ -267,25 +267,41 @@ int BitmapFont::width(const Utf8View& utf8) const
 {
     bool first = true;
     int width = 0;
+    int longest_width = 0;
 
     for (u32 code_point : utf8) {
+        if (code_point == '\n' || code_point == '\r') {
+            width = 0;
+            continue;
+        }
         if (!first)
             width += glyph_spacing();
         first = false;
         width += glyph_or_emoji_width(code_point);
+        if (width > longest_width)
+            longest_width = width;
     }
 
-    return width;
+    return longest_width;
 }
 
 int BitmapFont::width(const Utf32View& view) const
 {
     if (view.length() == 0)
         return 0;
     int width = (view.length() - 1) * glyph_spacing();
```

Comment:
This part might be wrong since I don't really understand why `width` is initialized with a non-zero value.

Reply:
You have the following situation:
```
[width of glyph 1][glyph spacing][width of glyph 2][glyph spacing][width of glyph 3]
```
This means you have `(length - 1) * spacing`, plus the width of all glyphs.
One idea to fix this would be to have a `size_t glyphs_on_line`, which you increment on every glyph and then reset when you encounter `\r` or `\n`. Then you can add the (glyphs_on_line - 1) * spacing` after the loop.

Reply:
ok so if I understand it correctly this diff should be ok?
```
diff --git a/Userland/Libraries/LibGfx/BitmapFont.cpp b/Userland/Libraries/LibGfx/BitmapFont.cpp
index b22b5c62d..32c09be2e 100644
--- a/Userland/Libraries/LibGfx/BitmapFont.cpp
+++ b/Userland/Libraries/LibGfx/BitmapFont.cpp
@@ -289,19 +289,19 @@ int BitmapFont::width(const Utf32View& view) const
 {
     if (view.length() == 0)
         return 0;
-    int width = (view.length() - 1) * glyph_spacing();
-    int longest_width = width;
+    int width = 0;
+    int longest_width = 0;
     for (size_t i = 0; i < view.length(); ++i) {
         auto code_point = view.code_points()[i];
         if (code_point == '\n' || code_point == '\r') {
-            width = (view.length() - 1) * glyph_spacing();
+            width = 0;
             continue;
         }
         width += glyph_or_emoji_width(code_point);
         if (width > longest_width)
             longest_width = width;
     }
-    return longest_width;
+    return (longest_width - 1) * glyph_spacing();
 }
 
 void BitmapFont::set_type(FontTypes type)
```

Reply:
No. I'll answer in a new comment, so I can make a proper `suggestion`.

Reply:
This should probably work.
```cpp
if (view.is_empty())
    return 0;
int longest_width = 0;
int glyphs_width = 0;
size_t glyph_count = 0;
for (auto code_point : view) {
    if (code_point == '\n' || code_point == '\r') {
        longest_width = max(longest_width, glyphs_width + (glyph_count - 1) * glyph_spacing());
        width = 0;
        glyph_count = 0;
        continue;
    }
    ++glyph_count;
    glyphs_width += glyph_or_emoji_width(code_point);
}
return longest_width;
```

Reply:
I wasn't exactly sure how to test it so I basically made Label pass Utf32View when calculating width and it seems like I'm either I'm testing it wrong or `BitmapFont::width(const Utf8View& utf8)` implementation works properly for Utf32View too.

```
diff --git a/Userland/Libraries/LibGUI/Label.cpp b/Userland/Libraries/LibGUI/Label.cpp
index 88a829bf8..05c7a8f34 100644
--- a/Userland/Libraries/LibGUI/Label.cpp
+++ b/Userland/Libraries/LibGUI/Label.cpp
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/Utf32View.h>
 #include <LibGUI/Label.h>
 #include <LibGUI/Painter.h>
 #include <LibGfx/Bitmap.h>
@@ -128,7 +129,11 @@ void Label::paint_event(PaintEvent& event)
 
 void Label::size_to_fit()
 {
-    set_fixed_width(font().width(m_text));
+    Vector<u32> v;
+    for (auto c : m_text)
+        v.append(c);
+
+    set_fixed_width(font().width(Utf32View { v.data(), v.size() }));
 }
 
 void Label::wrap_text()
```

With diff above plus your changes it seems like the width is calculated improperly:
![image](https://user-images.githubusercontent.com/6830979/124610048-7379d600-de70-11eb-87cf-49c8722c56f1.png)

Reply:
Sorry, I didn't look at the code close enough. Yes, you should be able to reuse the same code from `Utf8View` for `Utf32View`, since the differences are abstracted away when you iterate over them.